src/relayserver/RelayServer.ts

```diff
@@ -440,50 +439,53 @@ export class RelayServer extends EventEmitter {
-    const currentBlock = await this.contractInteractor.getBlockNumber()
-    const latestTxBlockNumber = await this._getLatestTxBlockNumber()
+  async getShouldRegisterAgain (currentBlock: number, hubEventsSinceLastScan: EventData[]): Promise<boolean> {
+    const latestTxBlockNumber = await this._getLatestTxBlockNumber(hubEventsSinceLastScan)
     return this.config.registrationBlockRate === 0 ? false : currentBlock - latestTxBlockNumber >= this.config.registrationBlockRate
   }
 
-  async handlePastHubEvents (blockNumber: number): Promise<void> {
+  async handlePastHubEvents (blockNumber: number, hubEventsSinceLastScan: EventData[]): Promise<void> {
+    for (const event of hubEventsSinceLastScan) {
+      switch (event.event) {
+        case TransactionRejectedByPaymaster:
+          log.debug('handle TransactionRejectedByPaymaster event', event)
+          await this._handleTransactionRejectedByPaymasterEvent(blockNumber)
+          break
+      }
+    }
+  }
+
+  async getAllHubEventsSinceLastScan (): Promise<EventData[]> {
     const topics = [address2topic(this.managerAddress)]
     const options = {
       fromBlock: this.lastScannedBlock + 1,
       toBlock: 'latest'
     }
-    const eventNames = [TransactionRejectedByPaymaster]
-    const decodedEvents = await this.contractInteractor.getPastEventsForHub(eventNames, topics, options)
-    for (const dlog of decodedEvents) {
-      switch (dlog.event) {
-        case TransactionRejectedByPaymaster:
-          await this._handleTransactionRejectedByPaymasterEvent(dlog, blockNumber)
-          break
-      }
-    }
+    return await this.contractInteractor.getPastEventsForHub(topics, options)
   }
 
-  async _handleTransactionRejectedByPaymasterEvent (dlog: EventData, blockNumber: number): Promise<void> {
-    log.debug('handle TransactionRejectedByPaymaster event', dlog)
+  async _handleTransactionRejectedByPaymasterEvent (blockNumber: number): Promise<void> {
     this.alerted = true
     this.alertedBlock = blockNumber
     console.error(`Relay entered alerted state. Block number: ${blockNumber}`)
   }
 
-  async _getLatestTxBlockNumber (): Promise<number> {
-    const events: EventData[] = await this.contractInteractor.getPastEventsForHub(constants.activeManagerEvents,
-      [address2topic(this.managerAddress)], {
-        fromBlock: 1
-      })
-    return events
-      .filter(
-        (e: EventData) =>
-          e.returnValues.relayManager.toLowerCase() === this.managerAddress.toLowerCase())
-      .map(
-        (e: EventData) =>
-          e.blockNumber)
-      .reduce(
-        (b1: any, b2: any) => Math.max(b1, b2), 0)
+  async _getLatestTxBlockNumber (eventsSinceLastScan: EventData[]): Promise<number> {
+    const latestTransactionSinceLastScan = reduceToLatestTx(eventsSinceLastScan)
+    if (latestTransactionSinceLastScan != null) {
+      this.lastMinedActiveTransaction = latestTransactionSinceLastScan
+    }
+    if (this.lastMinedActiveTransaction == null) {
```

Comment:
instead of checking each time if it is initialized, this is something that should be initialized once, during _init

Reply:
Note that there may be none if the server just started. This would be an awkward special case if I had to initialise `RelayServer` with the `lastMinedActiveTransaction` already not-null.